### DISCUSSION TITLE
Get rid of @trpc/server peer dependency?

### DISCUSSION THREAD

**Discussion by mmkal** [trpc/trpc#5753] [Ideas]
Something that would be really interesting would be to make `@trpc/client` (and the client wrapper libraries) _not_ have a peer dependency on `@trpc/server` anymore. Especially in v11, when procedures will have a $types key, it might be possible to define, in client-land, some types that _look like_ Router and Procedure types, without the dependency on the server library.

Simplified, non-edge case handling but functional types implementation that I think is roughly compatible with what a v11 server router looks like:

```ts
export type TRPCRouterLike = {
    _def: {
        procedures: Record<string, TRPCRouterLike | TRPCProcedureLike>
    }
}

export type TRPCProcedureLike = {
    _def: {
        $types: {
            input: any
            output: any
        }
    }
    procedure: true
    type: 'query' | 'mutation' | 'subscription'
}

export type TRPCClient<R extends TRPCRouterLike> = {
    [K in keyof R['_def']['procedures']]:
        R['_def']['procedures'][K] extends TRPCProcedureLike
            ? {
                [M in R['_def']['procedures'][K]['type']]: (input: inferProcedureInput<R['_def']['procedures'][K]>) => Promise<inferProcedureOutput<R['_def']['procedures'][K]>>
            }
            : R['_def']['procedures'][K] extends TRPCRouterLike
                ? TRPCClient<R['_def']['procedures'][K]>
                : never
}

export type inferProcedureInput<P extends TRPCProcedureLike> = P['_def']['$types']['input']
export type inferProcedureOutput<P extends TRPCProcedureLike> = P['_def']['$types']['output']

const createClient = <R extends TRPCRouterLike>() => {
    return {} as TRPCClient<R>
}
```

Then a router could either be a full, heavyweight thing, with all the baggage from the server (which is fine in a monopackage), or it could be a trimmed down typescript definition in a .d.ts file:

```ts
export type MyRouter: {
    _def: {
        procedures: {
            foo: {
                _def: {
                    $types: {
                        input: {status?: string}
                        output: {a: 123}
                    }
                }
                procedure: true
                type: 'query'
            }
            bar: {
                _def: {
                    procedures: {
                        abc: {
                            _def: {
                                $types: {
                                    input: {x: number}
                                    output: {a: 123}
                                }
                            }
                            procedure: true
                            type: 'mutation'
                        }
                    }
                }
            }
        }
    }
}

declare const myRouter: MyRouter
myRouter satisfies TRPCRouterLike

const myClient = createClient<MyRouter>()

myClient.foo.query satisfies (input: {status?: string}) => Promise<{a: 123}>

myClient.bar.abc.mutation satisfies (input: {x: number}) => Promise<{a: 123}>
```

([playground link including usage with the above hypothetical router](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBDAnmYcAqAlACgYQxAVxmCgBkBLAa1QF44BvAKDhbgH0ATYAMwC4HmrIWCgQAxsA4EowAM78MwMdA4AeWTCjkAdgHMANOmx5CxMlVQAfI7iyiJUmRWoA+QSwC+jL41CRYCMiomLb2ktLAzrQCQpw8-ExCQgAkSCjyMUlZOmBE-ACG2ojuWaymuTAFRSVCXrUlIuLhMvyaBMAlacD8AOQAju1QiD1w1j0AtkT5MOQQ2iNjsgQARrJiWmAzcz3ejL7g0PBdNjg4ADbkwNowqhhwoMTaHLIn+EQkUS5wdImsANoAaTgOjg1EQEG4cAwfx6cW4PQAujDGg4IrJEQjeDVWNDYVx4UieijmnJEYCEfcQI9nic7E1HJELNisgB+TKlUp-ACywO0UJhcLJRLCDPRSIBhK6GP4AAocnledwSHTUTIAJLaCq3AX4oXE0VkiUuACU3y+dPG5FkwFUOiVUBVJIA8kQtbjBYT9WjDQiXG4OfUA6wFDqeHqRd7xRSHlcaSETO9zNRmQG2fHzpdrtq8WHPRGZGLyf6g0l+NpgAA3Ei7fb+I5BRXK-PADVarCU6kveOOhmfb5wLChgkw1JBQs9eUwRG1w6BFCNh3Nl0wNsd2Nd4w9iJ9uiDnPDnqj9JC8pEaeMZTaDRwdbAabADNXeB0W5rp4b3BvMyfGWiRMKU0aC+X4WBkGBpD5egPDgfIP1OC4n1uNwfAAehQmCwDAPYuDEM58hkG85mvcZEC-EgEhKOEKI5L0C2oktuAgCB6JLVgqPZVikiPOQWM4jlJwSDRpgIWQWX4DQtD0Oo+IDU9KgYfJ+AARgAJgAZmkmTWE0zidNY2jugQKB2hTDkul6AYSGGUztJslhlnw3jWPYkCtLgAyMlctzWHyZYxCc7ykhcuzApYbjPJC0KhAEhgQDLAhxmWEg9KioM5ISRS4FUjTIsClLUvyqKDNaYyOlSrJzLgCYpi2eZcqDQrWMa0pms8FMdK8HwSLIqA4FkaYrW4S44J6qI9kvYjEEfa5+1ve9ppgGVutMEhjT2NC9hIhaADpGIgbbLKGPqBtkIa5DgOVNQVeghPA0TxM0HRdA8QDzVES1rVUehMuyjw3EYLaEOubaHKgbbfLEbbJhgAa5mOmZTuGi6YvoOK4G0BKkqgF6zQHd6rRtb7lPUv69iAA))

The above is pretty much what the `Procedure`, `inferProcedureInput` and `inferProcedureOutput` types already look like in the server package but the type for routers includes lots of information that clients shouldn't need to care about, like `ctx` types etc.

The nice thing about this is it might allow a tool like [xtrpc](https://github.com/algora-io/xtrpc) to output dramatically smaller `.d.ts` files for use by clients. And those `.d.ts` files would have no dependencies at all, they'd just be plain typescript. (Assuming it can reduce down zod types to their `z.infer<...>` results too).

Was this the intention with the addition of the `_def.$types` property to Procedures in v11? Maybe there's already a discussion about this somewhere, I haven't kept up with everything lately(!)

**Comment by juliusmarminge** [comment]
we tried this by adding a core lib that each package would have as a bundled dependency but had to drop due to `The inferred type of 'xxxx' cannot be named without a reference to '.pnpm/@trpc+core@[.....]'. This is likely not portable. A type annotation is necessary` (see archived Discord thread: https://discord.com/channels/867764511159091230/1196514174801551511)

We use the server package not only to share some types but also some runtime functions between packages, so just adding compatible types in `@trpc/client` would not be enough).

**Reply by mmkal** [reply]
Got it, yes that discussion rings a bell. That error haunts my dreams.

And thanks, yes I see some of the non-type @trpc/server imports in the client package.

I wonder whether separating out these problems might help:

1. compatible types in `@trpc/client` - no dependency of any kind for ensuring end-users provide valid routers/procedures
2. a `@trpc/server` (or `@trpc/server`) direct dependency for the runtime functions/internal types

But I do see this would be a significant change, `@trpc/server` is imported in the client package more than I thought it was. I'll take a look at https://github.com/trpc/tmp-trpc-declaration-repro and poke around a bit and report back if I come up with anything I think might be worth trying.
